### PR TITLE
Implemented API for Merkle proof

### PR DESCRIPTION
**Pull request by MichaelKim20**
Implemented to enable the Merkle proof in the network.
Register API Interface
Use Merkle Proof

MerkleTree_Proof

### CODE REVIEW

File: source/agora/node/Ledger.d [hunk 1]
```diff
@@ -202,6 +202,34 @@ public class Ledger
...
         return null;
     }
+
+
+    /***************************************************************************
+
+        Get the array of hashs the merkle branch.
+
+        Params:
+            block_height = the starting block height to begin retrieval from
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: Wrong documentation, it's for `getBlocksFrom`

File: source/agora/node/Ledger.d [hunk 2]
```diff
@@ -202,6 +202,34 @@ public class Ledger
...
+            hash         = transaction hash
+
+        Returns:
+            the array of hashs the merkle branch
+
+    ***************************************************************************/
+
+    public Hash[] getMerkleBranch (ulong block_height, Hash hash) @safe
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: This function should be unittested (not network tests but tests inside this module).

Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d]: I mean in addition to the network tests.

File: source/agora/common/API.d [hunk 3]
```diff
@@ -141,4 +141,24 @@ public interface API
     ***************************************************************************/
 
     public Block[] getBlocksFrom (ulong block_height, size_t max_blocks);
+
+
+    /***************************************************************************
+
+        Get the array of hashs the merkle branch.
```
Comment (AndrejMitrovic) [inline source/agora/common/API.d]: It's a bit hard to understand this documentation. Maybe to "get the array of hashes which form the merkle branch"? @Geod24 can you think of better wording?

File: source/agora/common/API.d [hunk 4]
```diff
@@ -141,4 +141,24 @@ public interface API
...
+            block_height = the starting block height to begin retrieval from
+            hash         = transaction hash
+
+        Returns:
+            the array of hashs the merkle branch
+
+    ***************************************************************************/
+
+    public Hash[] getMerkleBranch (ulong block_height, Hash hash);
+
```
Comment (AndrejMitrovic) [inline source/agora/common/API.d]:
```suggestion
```

File: source/agora/common/Block.d [hunk 5]
```diff
@@ -217,6 +217,26 @@ public struct Block
...
+
+        Params:
+            hash = `Hash` of `Transaction`
+
+        Returns:
+            Return sequence of transactions for the hash
+
+    *******************************************************************************/
+
+    public size_t findHashIndex(Hash hash) @safe
```
Comment (AndrejMitrovic) [inline source/agora/common/Block.d]:
```suggestion
    public size_t findHashIndex (Hash hash) @safe
```

File: source/agora/common/Block.d [hunk 6]
```diff
@@ -217,6 +217,26 @@ public struct Block
...
+
+    /*******************************************************************************
+
+        Find the sequence of transactions for the hash.
+
+        Params:
+            hash = `Hash` of `Transaction`
+
+        Returns:
+            Return sequence of transactions for the hash
```
Comment (AndrejMitrovic) [inline source/agora/common/Block.d]: It should document what it returns if the hash index is not found.

File: source/agora/common/API.d (hunk 3 above)
Comment (Geod24) [reply source/agora/common/API.d]: I assume "merkle branch" is actually a "merkle path", right ?

Comment (AndrejMitrovic) [reply source/agora/common/API.d]: In that case I suggest renaming the function to `getMerklePath`.

File: source/agora/node/Ledger.d (hunk 1 above)
Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: Okay, I'll fix it.

File: source/agora/node/Ledger.d (hunk 2 above)
Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: Okay

File: source/agora/common/API.d (hunk 4 above)
Comment (MichaelKim20) [reply source/agora/common/API.d]: Okay~

File: source/agora/common/Block.d (hunk 5 above)
Comment (MichaelKim20) [reply source/agora/common/Block.d]: Okay~

File: source/agora/common/Block.d (hunk 6 above)
Comment (MichaelKim20) [reply source/agora/common/Block.d]: Okay~

**Review by AndrejMitrovic — APPROVED** [review]

File: source/agora/common/Block.d [hunk 7]
```diff
@@ -332,39 +357,50 @@ unittest
...
-    const Hash h1 = hashFull(txs[1]);
-    const Hash h2 = hashFull(txs[2]);
-    const Hash h3 = hashFull(txs[3]);
-    const Hash h4 = hashFull(txs[4]);
-    const Hash h5 = hashFull(txs[5]);
-    const Hash h6 = hashFull(txs[6]);
-    const Hash h7 = hashFull(txs[7]);
+    Hash[] hashes;
+    hashes.length = txs.length;
+    hashes.reserve(txs.length);
```
Comment (AndrejMitrovic) [inline source/agora/common/Block.d:405]: Small suggestion:
```suggestion
```

The `.length` call above it has already allocated the memory space, so this `reserve` call doesn't do anything extra.

**Comment by AndrejMitrovic** [comment]
I think this is pretty much ready, unless @Geod24 has comments.

**Comment by Geod24** [comment]
Most of this PR is a rename from merkle branch to merkle path. It would have been much nicer / easier to make it a separate commit (and maybe even PR).
But it's a bit late in the sprint for that, so I will merge as-is. Let's make sure to split things more atomically next time.

**Review by Geod24 — APPROVED** [review]

**Pull request merged**